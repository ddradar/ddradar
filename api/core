C:/Users/ebise/Documents/Projects/ddradar-azure/core