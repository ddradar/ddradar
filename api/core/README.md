# @ddradar/core

Forked from [./core](../../core)
