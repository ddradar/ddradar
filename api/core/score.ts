import { StepChart } from './step-chart'

export type Score = Pick<StepChart, 'songId' | 'playStyle' | 'difficulty'> & {
  score: number
  exScore: number
  clearLamp: ClearLamp | null
  rank: DanceLevel | null
}

/**
 * 0: Failed
 * 1: Assisted Clear
 * 2: Clear
 * 3: Life 4 Clear
 * 4: (Good) Full Combo
 * 5: Great Full Combo
 * 6: Perfect Full Combo
 * 7: Marvelous Full Combo
 */
type ClearLamp = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7

type DanceLevel =
  | 'E'
  | 'D'
  | 'D+'
  | 'C-'
  | 'C'
  | 'C+'
  | 'B-'
  | 'B'
  | 'B+'
  | 'A-'
  | 'A'
  | 'A+'
  | 'AA-'
  | 'AA'
  | 'AA+'
  | 'AAA'

const isPositiveInteger = (num: number) => Number.isInteger(num) && num >= 0

export const getDanceLevel = (score: number): DanceLevel => {
  if (!isPositiveInteger(score) || score > 1000000)
    throw new RangeError('score is invalid')
  return score < 550000
    ? 'D'
    : score < 590000
    ? 'D+'
    : score < 600000
    ? 'C-'
    : score < 650000
    ? 'C'
    : score < 690000
    ? 'C+'
    : score < 700000
    ? 'B-'
    : score < 750000
    ? 'B'
    : score < 790000
    ? 'B+'
    : score < 800000
    ? 'A-'
    : score < 850000
    ? 'A'
    : score < 890000
    ? 'A+'
    : score < 900000
    ? 'AA-'
    : score < 950000
    ? 'AA'
    : score < 990000
    ? 'AA+'
    : 'AAA'
}

export const calcScore = (
  {
    notes,
    freezeArrow,
    shockArrow,
  }: Pick<StepChart, 'notes' | 'freezeArrow' | 'shockArrow'>,
  marvelous: number,
  perfect: number,
  great: number,
  good: number,
  ok: number
): number => {
  if (!isPositiveInteger(marvelous))
    throw new RangeError(`invalid marvelous: ${marvelous}`)
  if (!isPositiveInteger(perfect))
    throw new RangeError(`invalid perfect: ${perfect}`)
  if (!isPositiveInteger(great)) throw new RangeError(`invalid great: ${great}`)
  if (!isPositiveInteger(good)) throw new RangeError(`invalid good: ${good}`)
  if (!isPositiveInteger(ok)) throw new RangeError(`invalid ok: ${ok}`)
  if (marvelous + perfect + great + good > notes)
    throw new RangeError(
      `judge count(${
        marvelous + perfect + great + good
      }) is greater than notes(${notes})`
    )
  if (ok > freezeArrow + shockArrow)
    throw new RangeError(
      `ok count(${ok}) is greater than FA+SA count(${freezeArrow + shockArrow})`
    )

  const noteObjects = notes + freezeArrow + shockArrow
  const baseScore = 1000000 / noteObjects
  const rawScore =
    baseScore * (marvelous + ok) +
    (baseScore - 10) * perfect +
    (baseScore * 0.6 - 10) * great +
    (baseScore * 0.2 - 10) * good
  return Math.floor(rawScore / 10) * 10
}

export const calcExScore = (
  marvelous: number,
  perfect: number,
  great: number,
  ok: number
): number => {
  if (!isPositiveInteger(marvelous))
    throw new Error(`invalid marvelous: ${marvelous}`)
  if (!isPositiveInteger(perfect))
    throw new Error(`invalid perfect: ${perfect}`)
  if (!isPositiveInteger(great)) throw new Error(`invalid great: ${great}`)
  if (!isPositiveInteger(ok)) throw new Error(`invalid ok: ${ok}`)

  return marvelous * 3 + ok * 3 + perfect * 2 + great * 1
}
