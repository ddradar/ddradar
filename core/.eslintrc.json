{
  "root": true,
  "extends": ["../.eslintrc.json"],
  "rules": {
    "node/no-missing-import": "off"
  }
}
