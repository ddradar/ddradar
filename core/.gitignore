# Generated from staticwebapp.database.schema.gql via graphql-codegen
src/graphql.ts
