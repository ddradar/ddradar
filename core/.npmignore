# TypeScript Files
tsconfig.json
**.ts
!dist/**.d.ts

# Test & Lint Files
.eslintrc.json
jest.config.**
/__tests__/**
