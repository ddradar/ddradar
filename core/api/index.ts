export * from './auth'
export * from './course'
export * from './notification'
export * from './score'
export * from './song'
export * from './user'
