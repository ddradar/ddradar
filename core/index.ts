/* istanbul ignore file */

export * from './skill-attack'
export * from './typeUtils'
export * as Api from './api'
export * as Database from './db'
export * as Gate from './eagate-parser'
export * as Score from './score'
export * as Song from './song'
