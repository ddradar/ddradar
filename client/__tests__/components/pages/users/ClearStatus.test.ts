import type { Api } from '@ddradar/core'
import { Score } from '@ddradar/core'
import { mount } from '@vue/test-utils'

import Component from '~/components/pages/users/ClearStatus.vue'

describe('/components/pages/users/ClearStatus.vue', () => {
  describe('snapshot test', () => {
    test('renders correctly', async () => {
      // Arrange
      const title = 'LEVEL 10'
      const statuses: Pick<Api.ClearStatus, 'clearLamp' | 'count'>[] = [
        ...Array(9).keys(),
      ].map(i => ({
        clearLamp: (i - 1) as Score.ClearLamp | -1,
        count: 100 - i * 10,
      }))

      // Act
      const wrapper = mount(Component, { propsData: { title, statuses } })
      await wrapper.vm.$nextTick()

      // Assert
      const canvas = wrapper.element.getElementsByTagName('canvas')[0]
      const ctx = canvas.getContext('2d')
      expect(ctx).toMatchSnapshot()
    })
    test('renders empty', async () => {
      // Arrange - Act
      const wrapper = mount(Component)
      await wrapper.vm.$nextTick()

      // Assert
      const canvas = wrapper.element.getElementsByTagName('canvas')[0]
      const ctx = canvas.getContext('2d')
      expect(ctx).toMatchSnapshot()
    })
  })
})
