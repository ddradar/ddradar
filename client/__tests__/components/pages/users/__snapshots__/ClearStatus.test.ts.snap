// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/components/pages/users/ClearStatus.vue snapshot test renders correctly 1`] = `
Array [
  Object {
    "props": Object {
      "value": "12px \\"Helvetica Neue\\", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "value": "left",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "value": "middle",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
  Object {
    "props": Object {
      "text": "NoPlay",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  Object {
    "props": Object {
      "text": "Failed",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  Object {
    "props": Object {
      "text": "Assisted Clear",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  Object {
    "props": Object {
      "text": "Clear",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  Object {
    "props": Object {
      "text": "Life 4",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  Object {
    "props": Object {
      "text": "Full Combo",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  Object {
    "props": Object {
      "text": "Great Full Combo",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  Object {
    "props": Object {
      "text": "Perfect Full Combo",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  Object {
    "props": Object {
      "text": "Marvelous Full Combo",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
]
`;

exports[`/components/pages/users/ClearStatus.vue snapshot test renders empty 1`] = `
Array [
  Object {
    "props": Object {
      "value": "12px \\"Helvetica Neue\\", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "value": "left",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "value": "middle",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
]
`;
